### ISSUE TITLE
Asynchronous intercept method of StreamInterceptor

### ISSUE THREAD

**Issue by alienvspredator** [grpc/grpc-web#956]
Is it possible to implement promise- or callback- based `intercept` method of `StreamInterceptor` like the `UnaryInterceptor`? For example:

```ts
class StreamAuthInterceptor<REQ, RESP> implements StreamInterceptor<REQ, RESP> {
	async intercept(
		r: Request<REQ, RESP>,
		invoker: (r: Request<REQ, RESP>) => ClientReadableStream<RESP>
	): Promise<ClientReadableStream<RESP>> {
		r.getMetadata()['authorization'] = await getAuth();
		return invoker(r);
	}
}
```

**Comment by dhax** [comment]
did you find a solution to this? Having the same issue trying to await for a jwt token refresh...

**Comment by kurnal-volt** [comment]
@alienvspredator @dhax any solutions? running into the same use case also with doing JWT token refresh in the interceptor

**Comment by dhax** [comment]
@kurnal-volt no, I am only using an unaryInterceptor with the promise based client. Doesn't look like there was any progress regarding an async streamInterceptor.

**Comment by kurnal-volt** [comment]
@dhax could you show me what an example for a promise based unaryInterceptor is? also is there a promise based streamInterceptor?
